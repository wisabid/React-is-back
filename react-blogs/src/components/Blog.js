import React from 'react';
import PropTypes from 'prop-types';
import blogWrapper from './../HOC/blogWrapper'

const displayDate = (dt) => {
    return new Date(dt).toString();
}

const ReadMore = (props) => {
    if (props.desc.length > 80) {
        return (
            <>{props.desc.substr(0,80)}<span className="tiny-font" onClick={props.readmore}><i>...Read More</i></span></>
        )
    }
    else {
        return (
            props.desc
        )
    }
   
}


const Blog = (props) => {
    let blogheader,
        blogbody;

    function handleBodyFocus(ev) {
        if (ev.keyCode === 13) {
            blogbody.focus();
        }
    }

    if (props.edit) {
        return (
            <>
                { props.msgdisplay
                    ? <div className="blog-item error-container" style={props.styles}>{props.msgdisplay}</div>
                    : null
                } 
                <div className="blog-item highight edit-item active">
                    <div className="float-right tiny-font">
                        <span id="cancelBtn" onClick={props.cancelEdit}>Cancel</span> | <span id="updateBtn" onClick={props.updateBlog}>Update</span>
                    </div>
                    <input name="blogheader" type="text" value={props.header} size="115" onChange={props.bloghdr}
                    ref={(input) => blogheader = input} onKeyUp={handleBodyFocus} maxLength="200"/>
                    <br />
                    <textarea name="blogbody" cols="104" rows="10" value={props.body} onChange={props.blogbdy} 
                    ref={(input) => blogbody = input}></textarea>
                </div>
            </>
        )
    }
    else if (props.addblog) {
        return (
            <>
                { props.msgdisplay
                    ? <div className="blog-item error-container" style={props.styles}>{props.msgdisplay}</div>
                    : null
                }             
                <div className="blog-item">
                    <div className="float-right tiny-font">
                        <span id="cancelBtn" onClick={props.cancelAdd}>Cancel</span> | <span id="submitbtn" onClick={props.submitblog}>Submit</span>
                    </div>
                    <input ref={(input) => blogheader = input} placeholder="Header goes here" name="blogheader" type="text" size="104" onChange={props.bloghdr} 
                    onKeyUp={handleBodyFocus} maxLength="200" onBlur={props.similar}/>
                    <br />
                    <textarea ref={(input) => blogbody = input} placeholder="Body goes here" name="blogbody" cols="95" rows="10" onChange={props.blogbdy}></textarea>
                </div>
            </>
        )
    }
    else {
        return (
            <>
                { props.msgdisplay
                    ? <div className="blog-item error-container" style={props.styles}>{props.msgdisplay}</div>
                    : null
                }                 
                <div className="blog-item" >
                    <div className="float-right tiny-font">
                        <span className="editBtn" onClick={props.editBlog}>Edit</span> | <span className="deleteBtn" onClick={props.deleteBlog}>Delete</span>
                    </div>
                    <h3>{props.header}</h3>
                    <span className="tiny-font" style={{color:'darkseagreen', cursor: 'crosshair'}}>({displayDate(props.postdate)})</span>
                    <p><pre>{ (props.body.length > 80 && !props.showfull)? <ReadMore desc={props.body} readmore={props.entirebody}/>:props.body}</pre></p>
                </div>
            </>
        )
    }
}

Blog.propTypes = {
    header : PropTypes.string.isRequired,
    body: PropTypes.string.isRequired
};

export default React.memo(blogWrapper(Blog));