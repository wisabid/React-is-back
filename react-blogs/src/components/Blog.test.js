import React from 'react';
import Blog from './Blog';
import { shallow, mount } from 'enzyme';

describe('<Blog />', () => {
    const clickFn = jest.fn();
  
    it('renders Blog component without crashing', () => {
      const component = shallow(<Blog addblog={true} suggestions={[]}/>);
      expect(component).toMatchSnapshot();
    })
  
    it('Add New link should display an input element for adding up a new blog', () => {
      const wrapper = shallow(<Blog addblog={true} suggestions={[]}/>);
      expect(wrapper.find('input[name="blogheader"]')).toHaveLength(1);    
    })

    it('should be possible to open Add new blog form', () => {
      const component = mount(<Blog addblog={true} suggestions={[]}/>);
      component
        .find('span#submitbtn')
        .simulate('click');
      expect(component).toMatchSnapshot();
      component.unmount();
    });
  
    it('Add New link should display an textarea element for adding up a new blog', () => {
      const wrapper = shallow(<Blog addblog={true} suggestions={[]}/>);
      expect(wrapper.find('textarea[name="blogbody"]')).toHaveLength(1);    
    })
  
    it('Add blog should have a Submit button for adding up', () => {
      const wrapper = shallow(<Blog addblog={true} suggestions={[]}/>);
      expect(wrapper.find('span#submitbtn')).toHaveLength(1);    
    })
  
    it('Submit click should call a function to submit', () => {
      const wrapper = shallow(<Blog addblog={true} submitblog={clickFn} suggestions={[]}/>);
      wrapper.find('span#submitbtn').simulate('click');
      expect(clickFn).toHaveBeenCalled();
    })
  
    it('Cancel click should call a function to cancel', () => {
      const wrapper = shallow(<Blog addblog={true} cancelAdd={clickFn} suggestions={[]}/>);
      wrapper.find('span#cancelBtn').simulate('click');
      expect(clickFn).toHaveBeenCalled();
    })
  
    // it('Edit link should have been called with an id', () => {
    //   const wrapper = shallow(<Blog edit={true} editBlog={() => clickFn('abcd_efgh')} header="Blog header" body="Blog body"/>);
    //   wrapper.find('span#editBtn').simulate('click');
    //   expect(clickFn).toHaveBeenCalledWith('abcd_efgh');    
    // })

    it('Edit link should be able to update an existing blog', () => {
      const component = mount(<Blog edit={true}/>);
      component
        .find('span#updateBtn')
        .simulate('click');
      expect(component).toMatchSnapshot();
      component.unmount();
    });
  
    it('Edit link should edit a listed blog', () => {
      const wrapper = shallow(<Blog edit={true}/>);
      expect(wrapper.find('.blog-item.edit-item')).toHaveLength(1);    
    })
  
    it('Edit blog should display a textbox element for editing a new blog', () => {
      const wrapper = shallow(<Blog edit={true}/>);
      expect(wrapper.find('textarea[name="blogbody"]')).toHaveLength(1);    
    })
  
    it('Edit blog should display a textarea element for editing a new blog', () => {
      const wrapper = shallow(<Blog edit={true}/>);
      expect(wrapper.find('input[name="blogheader"]')).toHaveLength(1);    
    })
  
    it('Edit blog should have an Update button for updating', () => {
      const wrapper = shallow(<Blog edit={true}/>);
      expect(wrapper.find('span#updateBtn')).toHaveLength(1);    
    })
  
    it('Update click should call a function to update', () => {
      const wrapper = shallow(<Blog edit={true} editBlog={clickFn} />);
      wrapper.find('span#updateBtn').simulate('click');
      expect(clickFn).toHaveBeenCalled();
    })
    
    it('Cancel click while editing should call a function to cancel', () => {
      const wrapper = shallow(<Blog edit={true} cancelEdit={clickFn} />);
      wrapper.find('span#cancelBtn').simulate('click');
      expect(clickFn).toHaveBeenCalled();
    })
  
    it('Listing page should display a blog when header and body are passed', () => {
      const wrapper = shallow(<Blog header="Blog header" body="Blog body"/>);
      expect(wrapper.find('.blog-item')).toHaveLength(1);    
    })
  
    it('A blog listing should have a delete button', () => {
      const wrapper = shallow(<Blog header="Blog header" body="Blog body"/>);
      expect(wrapper.find('span.deleteBtn')).toHaveLength(1);    
    })
  
    it('A blog listing should have an edit button', () => {
      const wrapper = shallow(<Blog header="Blog header" body="Blog body"/>);
      expect(wrapper.find('span.editBtn')).toHaveLength(1);    
    })
  
    it('A blog body should limit to 80 characters display in listing page in addition to REad more link', () => {
      const wrapper = shallow(<Blog header="Blog header" body="Iam a loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong Blog body"/>);
      expect(wrapper.find('.blog-item p pre').html()).toContain('<span class="tiny-font"><i>...Read More</i>');    
    })
  
    it('A click on Read more link should call a function', () => {
      const wrapper = shallow(<Blog showfull={clickFn} header="Blog header" body="Iam a loooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooong Blog body"/>);
      wrapper.find('.blog-item .tiny-font').at(0).simulate('click');
      expect(clickFn).toHaveBeenCalled();
    })
  
  })