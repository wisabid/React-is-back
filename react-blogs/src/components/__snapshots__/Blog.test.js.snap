// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Blog /> Edit link should be able to update an existing blog 1`] = `
ReactWrapper {
  Symbol(enzyme.__unrendered__): <UNDEFINED
    edit={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "edit": true,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": <div
          class="float-right tiny-font"
        >
          <span
            id="cancelBtn"
          >
            Cancel
          </span>
           | 
          <span
            id="updateBtn"
          >
            Update
          </span>
        </div>,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              id="cancelBtn"
            >
              Cancel
            </span>,
            " | ",
            <span
              id="updateBtn"
            >
              Update
            </span>,
          ],
          "className": "float-right tiny-font",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": <span
              id="cancelBtn"
            >
              Cancel
            </span>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cancel",
              "id": "cancelBtn",
              "onClick": undefined,
            },
            "ref": null,
            "rendered": Array [
              "Cancel",
            ],
            "type": "span",
          },
          " | ",
          Object {
            "instance": <span
              id="updateBtn"
            >
              Update
            </span>,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Update",
              "id": "updateBtn",
              "onClick": undefined,
            },
            "ref": null,
            "rendered": Array [
              "Update",
            ],
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": <input
          maxlength="200"
          name="blogheader"
          size="115"
          type="text"
          value=""
        />,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "maxLength": "200",
          "name": "blogheader",
          "onChange": undefined,
          "onKeyUp": [Function],
          "size": "115",
          "type": "text",
          "value": undefined,
        },
        "ref": [Function],
        "rendered": Array [
          undefined,
        ],
        "type": "input",
      },
      Object {
        "instance": <br />,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": Array [
          undefined,
        ],
        "type": "br",
      },
      Object {
        "instance": <textarea
          cols="104"
          name="blogbody"
          rows="10"
        />,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "cols": "104",
          "name": "blogbody",
          "onChange": undefined,
          "rows": "10",
          "value": undefined,
        },
        "ref": [Function],
        "rendered": Array [
          undefined,
        ],
        "type": "textarea",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "edit": true,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": <div
            class="float-right tiny-font"
          >
            <span
              id="cancelBtn"
            >
              Cancel
            </span>
             | 
            <span
              id="updateBtn"
            >
              Update
            </span>
          </div>,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                id="cancelBtn"
              >
                Cancel
              </span>,
              " | ",
              <span
                id="updateBtn"
              >
                Update
              </span>,
            ],
            "className": "float-right tiny-font",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": <span
                id="cancelBtn"
              >
                Cancel
              </span>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Cancel",
                "id": "cancelBtn",
                "onClick": undefined,
              },
              "ref": null,
              "rendered": Array [
                "Cancel",
              ],
              "type": "span",
            },
            " | ",
            Object {
              "instance": <span
                id="updateBtn"
              >
                Update
              </span>,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Update",
                "id": "updateBtn",
                "onClick": undefined,
              },
              "ref": null,
              "rendered": Array [
                "Update",
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": <input
            maxlength="200"
            name="blogheader"
            size="115"
            type="text"
            value=""
          />,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "maxLength": "200",
            "name": "blogheader",
            "onChange": undefined,
            "onKeyUp": [Function],
            "size": "115",
            "type": "text",
            "value": undefined,
          },
          "ref": [Function],
          "rendered": Array [
            undefined,
          ],
          "type": "input",
        },
        Object {
          "instance": <br />,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": Array [
            undefined,
          ],
          "type": "br",
        },
        Object {
          "instance": <textarea
            cols="104"
            name="blogbody"
            rows="10"
          />,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "cols": "104",
            "name": "blogbody",
            "onChange": undefined,
            "rows": "10",
            "value": undefined,
          },
          "ref": [Function],
          "rendered": Array [
            undefined,
          ],
          "type": "textarea",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
